fdbclient/StorageServerInterface.h

```diff
@@ -278,11 +283,35 @@ struct GetKeyValuesReply : public LoadBalancedReply {
 	}
 };
 
+// Describe how to hop. Assume the keys are encoded as Tuple, hence there are concept of "elements".
+struct HopInfo {
+	// Do GetKeyValues, extract the last suffixLen elements from each keys.
+	// TODO: If it is 0, use the entire value rather part of the key.
+	int suffixLen;
+	// Form hop keys by appending the fetched suffixes to hopPrefix. Do GetValue using the hop keys.
+	// TODO: Support getting ranges using the hop keys as prefixes.
+	KeyRef hopPrefix;
```

Comment:
I think this covers the important cases, but to make the scope explicit. 

The simplest case is where the scanned keys are
```
dir, 'I', value, id1, id2
```
and the key-values to be fetched are
```
dir, 'R', id1, id2, n
```
(`n` might be a split record counter or an exploded per-field key)

This also handles cases where the scanned keys are
```
dir, 'I', id1, value, id2
```
and the key-values to be fetched are still
```
dir, 'R', id1, id2, n
```
provided that the scan fixes `id1` so that it can be included in the skip prefix and the hop prefix. That is,
```sql
WHERE id1 = ? AND value BETWEEN ? AND ? ORDER BY  value
```

It is not sufficient for cases where `id1` needs to be extracted for the hop. Such as,
```sql
WHERE id1 > ? ORDER BY id1, value
```

Nor for the case where the index is
```
dir, 'I', id2, id1
```
for scans like
```sql
id2 > ? ORDER BY id2
```

Again, I am not sure more complicated tuple element shuffling descriptors are worth the trouble. The two supported cases match the ones I have seen in real life.

Reply:
Yeah, I can update `int suffixLen` to a vector of picked item indexes in the expected order.

To make it more expressive, and probably more elegant, I'm also thinking about using a single tuple to express both `suffixLen` and `hopPrefix`. It could be something like:
```
dir, "R", "%2","%4", "%.*"
```
"%.*" is a language sugar representing doing a range query rather than point query.
Escaping on the strings are needed, of course.

Is there a use case of interleaving the fetched items with the constant items?

Reply:
> Is there a use case of interleaving the fetched items with the constant items?

We might distinguish three kinds of hop elements:
* variable, fetched from the scanned keys
* constant, occurring in both the scan range and the hop prefix
* constant, occurring only in the hop prefix

The scanned range might arbitrarily permute its elements for the sake of ordering or inequality comparisons. So the first two interleave arbitrarily, too.

Placing the last someplace other than at the front is maybe a little more contrived, but suppose you have
```
dir, "I", value, group, id
```
and
```
dir, "R", group, "widget", id
```
where the constant type (_widget_) is implied for the original scan -- it's an index on that type only -- and so not part of its keys, but the typing is normalized within some form of group.